### ISSUE TITLE
`unused-ignore-comment` in unreachable code

### ISSUE THREAD

**Issue by pawamoy** [astral-sh/ty#2681]
### Summary

I'm checking code on different Python versions.

```python
if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib
```

On Python 3.10, `tomli` is installed. On Python 3.11 and higher, it is not, so ty will report an `unresolved-import`. Maybe later ty will understand the `if sys.version_info ...` condition above, but in the meantime I just want to ignore it, so I add a ty-ignore comment:

```python
if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib  # ty: ignore[unresolved-import]
```

However, it now fails with `unused-ignore-comment` on Python 3.10. Fair enough, I ignore that too:

```python
if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib  # ty: ignore[unresolved-import, unused-ignore-comment]
```

But now it fails with `unused-ignore-comment` on Python 3.11 and higher. I noticed that I can *stack* these comments:

```python
if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib  # ty: ignore[unresolved-import, unused-ignore-comment, unused-ignore-comment]
```

😂 

But shouldn't just one of them be enough?

### Version

0.0.14

**Comment by carljm** [comment]
Just to clarify some things (which I think are not related to your core point):

> Maybe later ty will understand the `if sys.version_info ...` condition above

ty definitely understands `if sys.version_info` conditions, and has for a year already.

> On Python 3.11 and higher, it is not, so ty will report an `unresolved-import`

ty should not report any error on this code if you are checking with Python 3.11 or higher, and in my test it does not:

```
➜ cat main.py
import sys

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

➜ uvx ty check . --python-version 3.11
All checks passed!
```

Checking this code on Python 3.10 does emit an error, if I don't have a venv with tomli installed in it:

```
➜ uvx ty check . --python-version 3.10 --output-format concise
main.py:6:12: error[unresolved-import] Cannot resolve imported module `tomli`
Found 1 diagnostic
```

So maybe this is the reason you need to suppress `unresolved-import`?

I think the best solution for this kind of situation is to use a Python 3.10 venv (with tomli installed) when checking your codebase under 3.10 -- then no suppression comments are required here at all.

But I think I also agree that if we are already suppressing `unused-ignore-comment`, we probably shouldn't emit `unused-ignore-comment` about the unused ignore of `unused-ignore-comment` :)

**Comment by pawamoy** [comment]
I do use a different venv for each check! I'll double check everything on my end and try to provide an MRE 🙂 Thanks!

**Comment by pawamoy** [comment]
From `ty check --help`:

> If a version is not specified on the command line or in a configuration file, ty will try the following techniques in order of preference to determine a value:
> 1. Check for the `project.requires-python` setting in a `pyproject.toml` file and use the minimum version from the specified range
> 2. Check for an activated or configured Python environment and attempt to infer the Python version of that environment
> 3. Fall back to the latest stable Python version supported by ty (see `ty check --help` output)

I have `requires-python = ">=3.10"` in my pyproject.toml, so whatever the virtualenv currently activated, ty will always check for 3.10, no?

UPDATE: yup seems like it:

```console
% for py in 3.10 3.11 3.12 3.13 3.14 3.15; do make $py python -V; make $py ty check -vvv 2>&1 | grep 'INFO ty_project::metadata::options:'; done
Python 3.10.19
  2026-02-02T13:17:01.743652Z  INFO ty_project::metadata::options: Defaulting to python-platform `linux`
  2026-02-02T13:17:01.744030Z  INFO ty_project::metadata::options: Python version: Python 3.10, platform: linux
Python 3.11.14
  2026-02-02T13:17:02.004054Z  INFO ty_project::metadata::options: Defaulting to python-platform `linux`
  2026-02-02T13:17:02.004428Z  INFO ty_project::metadata::options: Python version: Python 3.10, platform: linux
Python 3.12.12
  2026-02-02T13:17:02.261615Z  INFO ty_project::metadata::options: Defaulting to python-platform `linux`
  2026-02-02T13:17:02.261995Z  INFO ty_project::metadata::options: Python version: Python 3.10, platform: linux
Python 3.13.11
  2026-02-02T13:17:02.543691Z  INFO ty_project::metadata::options: Defaulting to python-platform `linux`
  2026-02-02T13:17:02.544062Z  INFO ty_project::metadata::options: Python version: Python 3.10, platform: linux
Python 3.14.2
  2026-02-02T13:17:02.795028Z  INFO ty_project::metadata::options: Defaulting to python-platform `linux`
  2026-02-02T13:17:02.795406Z  INFO ty_project::metadata::options: Python version: Python 3.10, platform: linux
Python 3.15.0a3
  2026-02-02T13:17:03.075537Z  INFO ty_project::metadata::options: Defaulting to python-platform `linux`
  2026-02-02T13:17:03.075917Z  INFO ty_project::metadata::options: Python version: Python 3.10, platform: linux
```

**Comment by pawamoy** [comment]
Passing an explicit `--python-version` fixes the issue for me, though the issue might still be relevant, as you pointed out in your last sentence 🙂  (also maybe it doesn't make sense for me to type-check for versions higher than my minimum required version 3.10, I only run lints/checks on every supported Python versions to make sure everything works for contributors, whatever the supported version they use locally).

> if we are already suppressing unused-ignore-comment, we probably shouldn't emit unused-ignore-comment about the unused ignore of unused-ignore-comment 

I'll let you decide if it should be closed or not. Thanks again!

**Comment by carljm** [comment]
Yeah, ty doesn't use the venv Python version, though this is a debatable choice and we could reconsider it.

I'll leave this issue open for now to track the specific idea of never emitting unused-ignore about an unused-ignore suppression comment.

**Comment by AlexWaygood** [comment]
> Yeah, ty doesn't use the venv Python version

(We do, but only as a fallback if you don't have a version set in a config file, which includes setting the `project.requires-python` field in a `pyproject.toml` file)

**Comment by MichaReiser** [comment]
Having to use two `unused-ignore-comment` suppressions feels silly, but there's value in warning about `unused-ignore-comment`s in general. 

Clippy (Rust linter) has a similar problem with conditionally compiled code and the use of `#[expect(...)]` (which requires that the code contain a given lint). The compilation will fail if the code only includes the code for some targets/features. 

A compromise could be to mark all `ignore-comment`s as used if we detect an unreachable branch. This would remove the need for adding any `unused-ignore-comment`s. But I'm not a 100% sure if this will cause any issues with suppressing unreachable/dead-code diagnostics. So I'd prefer to wait to make a change here until we have some dead code/unreachable diagnostics implemented, unless this comes up more often.

**Comment by alex** [comment]
FWIW, https://play.ty.dev/f5f0bade-1ac0-4a0d-9e45-4fafa2bbbf07 is a nice self-contained reproducer that can be made to work between arbitrary versions.
